File: EmotionEngine/src/main/java/co/q64/emotion/util/Rational.java
package co.q64.emotion.util;

import java.math.BigInteger;

public final class Rational implements Comparable<Rational> {
    public final static Rational ZERO = of(0);
    public final static Rational ONE = of(1);
    public final static Rational MINUS_ONE = of(-1);

    private BigInteger numerator;
    private BigInteger denominator;

    private Rational(BigInteger numerator, BigInteger denominator) {
        if (denominator.equals(BigInteger.ZERO))
            throw new ArithmeticException();

        this.numerator = numerator;
        this.denominator = denominator;

        reduce();
    }

    private Rational(BigInteger numerator) {
        this(numerator, BigInteger.ONE);
    }

    private Rational() {
        this(BigInteger.ZERO);
    }

    private void reduce() {
        boolean sign = false;
        if (numerator.signum() >= 0 ^ denominator.signum() >= 0) {
            sign = true;
        } else {
            if (numerator.equals(BigInteger.ZERO)) {
                denominator = BigInteger.ONE;
            }
        }

        numerator = numerator.abs();
        denominator = denominator.abs();

        if (sign) {
            numerator = numerator.multiply(BigInteger.valueOf(-1));
        }

        final BigInteger divisor = pgcd(numerator.abs(), denominator);
        if (!divisor.equals(BigInteger.ONE)) {
            numerator = numerator.divide(divisor);
            denominator = denominator.divide(divisor);
        }
    }

    public int intValue() {
        return numerator.intValue() / denominator.intValue();
    }

    public long longValue() {
        return numerator.longValue() / denominator.longValue();
    }

    public float floatValue() {
        return numerator.floatValue() / denominator.floatValue();
    }

    public double doubleValue() {
        return numerator.doubleValue() / denominator.doubleValue();
    }

    public static Rational of(long num, long denom) {
        return new Rational(BigInteger.valueOf(num), BigInteger.valueOf(denom));
    }

    public static Rational of(long num) {
        return of(num, 1);
    }

    public static Rational of(String s) {
        if (s.contains(".")) {
            int digitsDec = s.length() - 1 - s.indexOf('.');
            String d = "1";
            for (int i = 0; i < digitsDec; i++) {
                d += "0";
            }
            return of(s.replace(".", ""), d);
        }
        return of(s, "1");
    }

    public static Rational of(double d) {
        return of(String.valueOf(d));
    }

    public static Rational of(String n, String d) {
        return new Rational(new BigInteger(n), new BigInteger(d));
    }

    public BigInteger getNumerator() {
        return numerator;
    }

    public BigInteger getDenominator() {
        return denominator;
    }

    public Rational add(Rational n) {
        BigInteger num1 = numerator, denom1 = denominator, num2 = n.numerator, denom2 = n.denominator;

        num1 = num1.multiply(denom2);
        num2 = num2.multiply(denom1);
        BigInteger denom = denom1.multiply(denom2);

        return new Rational(num1.add(num2), denom);
    }

    public Rational substract(Rational n) {
        return add(new Rational(n.numerator.multiply(BigInteger.valueOf(-1)), n.denominator));
    }

    public Rational multiply(Rational n) {
        BigInteger num1 = numerator, denom1 = denominator, num2 = n.numerator, denom2 = n.denominator;

        num1 = num1.multiply(num2);
        denom1 = denom1.multiply(denom2);

        return new Rational(num1, denom1);
    }

    public Rational divide(Rational n) {
        return multiply(new Rational(n.denominator, n.numerator));
    }

    public Rational reciprocal() {
        return ONE.divide(this);
    }

    public boolean isInteger() {
        return denominator.equals(BigInteger.ONE);
    }

    public boolean isOne() {
        return numerator.equals(denominator);
    }

    public Rational abs() {
        return new Rational(numerator.abs(), denominator);
    }

    public Rational pow(int n) {
        Rational b = ONE;

        for (int i = 0; i < n; i++) {
            b = b.multiply(this);
        }

        return b;
    }

    private BigInteger pgcd(BigInteger a, BigInteger b) {
        BigInteger r;
        while (!b.equals(BigInteger.ZERO)) {
            r = a.mod(b);
            a = b;
            b = r;
        }
        return a;
    }

    @Override
    public String toString() {
        if (denominator.longValue() == 1) {
            return numerator.toString();
        }
        return numerator.toString() + "/" + denominator.toString();
    }

    @Override
    public int hashCode() {
        return toString().hashCode(); // Since there is one unique string representation, this should be unique too.
    }

    @Override
    public boolean equals(Object object) {
        if (object instanceof Rational) {
            Rational n = (Rational) object;
            if (numerator.equals(n.numerator) && denominator.equals(n.denominator)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public int compareTo(Rational o) {
        BigInteger a = numerator.multiply(o.denominator);
        BigInteger b = denominator.multiply(o.numerator);

        return a.compareTo(b);
    }
}

File: EmotionEngine/src/main/kotlin/co/q64/emotion/math/Rational.kt
package co.q64.emotion.math

import java.math.BigInteger


data class Rational(private var n: BigInteger = BigInteger.ZERO, private var d: BigInteger = BigInteger.ONE) : Comparable<Rational> {
    private val numerator: BigInteger
    private val denominator: BigInteger

    constructor(n: Long, d: Long = 1) : this(BigInteger.valueOf(n), BigInteger.valueOf(d))
    constructor(n: Int, d: Int = 1) : this(n.toLong(), d.toLong())
    constructor(n: String, d: String) : this(BigInteger(n), BigInteger(d))
    constructor(d: Double) : this(d.toString())
    constructor(s: String) : this(
            s.replace(".", ""),
            if (s.contains(".")) {
                val digitsDec = s.length - 1 - s.indexOf('.')
                var d = "1"
                for (i in 0 until digitsDec) {
                    d += "0"
                }
                d
            } else "1"
    )

    init {
        // TODO refactor and remove var
        var sign = false
        if (n.signum() >= 0 || d.signum() >= 0) {
            sign = true
        } else {
            if (n == BigInteger.ZERO) {
                d = BigInteger.ONE
            }
        }
        n = n.abs()
        d = d.abs()
        if (sign) {
            n = n.multiply(BigInteger.valueOf(-1))
        }
        val divisor = pgcd(n.abs(), d)
        if (divisor != BigInteger.ONE) {
            numerator = n.divide(divisor)
            denominator = d.divide(divisor)
        } else {
            numerator = n
            denominator = d
        }
    }

    fun intValue(): Int = numerator.toInt() / denominator.toInt()
    fun longValue(): Long = numerator.toLong() / denominator.toLong()
    fun floatValue(): Float = numerator.toFloat() / denominator.toFloat()
    fun doubleValue(): Double = numerator.toDouble() / denominator.toDouble()

    fun add(n: Rational): Rational {
        var num1 = numerator
        val denom1 = denominator
        var num2 = n.numerator
        val denom2 = n.denominator
        num1 = num1.multiply(denom2)
        num2 = num2.multiply(denom1)
        val denom = denom1.multiply(denom2)
        return Rational(num1.add(num2), denom)
    }

    fun subtract(n: Rational): Rational {
        return add(Rational(n.numerator.multiply(BigInteger.valueOf(-1)), n.denominator))
    }

    fun multiply(n: Rational): Rational {
        var num1 = numerator
        var denom1 = denominator
        val num2 = n.numerator
        val denom2 = n.denominator
        num1 = num1.multiply(num2)
        denom1 = denom1.multiply(denom2)
        return Rational(num1, denom1)
    }

    fun divide(n: Rational): Rational = multiply(Rational(n.denominator, n.numerator))
    fun reciprocal(): Rational = ONE.divide(this)
    val isInteger: Boolean get() = denominator == BigInteger.ONE
    val isOne: Boolean get() = numerator == denominator
    fun abs(): Rational = Rational(numerator.abs(), denominator)


    fun pow(n: Int): Rational {
        var b = ONE
        for (i in 0 until n) {
            b = b.multiply(this)
        }
        return b
    }

    private fun pgcd(a: BigInteger, b: BigInteger): BigInteger {
        var a1 = a
        var b1 = b
        var r: BigInteger
        while (b1 != BigInteger.ZERO) {
            r = a1.mod(b1)
            a1 = b1
            b1 = r
        }
        return a1
    }

    override fun toString(): String {
        return if (denominator.toLong() == 1L) {
            numerator.toString()
        } else "$numerator/$denominator"
    }

    override operator fun compareTo(other: Rational): Int {
        val a = numerator.multiply(other.denominator)
        val b = denominator.multiply(other.numerator)
        return a.compareTo(b)
    }

    companion object {
        val ZERO = Rational(0)
        val ONE = Rational(1)
        val MINUS_ONE = Rational(-1)
    }
}

fun Int.rational() = Rational(this)
fun Long.rational() = Rational(this)
fun String.rational() = Rational(this)
fun Double.rational() = Rational(this)
fun Boolean.rational() = if (this) Rational.ONE else Rational.ZERO